apps/obsidian/src/components/ModifyNodeModal.tsx

```diff
@@ -0,0 +1,515 @@
+
+type ModifyNodeFormProps = {
+  nodeTypes: DiscourseNode[];
+  onSubmit: (params: {
+    nodeType: DiscourseNode;
+    title: string;
+    initialFile?: TFile; // for edit mode
+    selectedExistingNode?: TFile;
+  }) => Promise<void>;
+  onCancel: () => void;
+  initialTitle?: string;
+  initialNodeType?: DiscourseNode;
+  initialFile?: TFile; // for edit mode
+  plugin: DiscourseGraphPlugin;
+};
+
+export const ModifyNodeForm = ({
+  nodeTypes,
+  onSubmit,
+  onCancel,
+  initialTitle = "",
+  initialNodeType,
+  initialFile,
+  plugin,
+}: ModifyNodeFormProps) => {
+  const isEditMode = !!initialFile;
+  const [title, setTitle] = useState(initialFile?.basename || initialTitle);
+  const [selectedNodeType, setSelectedNodeType] =
+    useState<DiscourseNode | null>(initialNodeType || null);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [selectedExistingNode, setSelectedExistingNode] =
+    useState<TFile | null>(null);
+  const [query, setQuery] = useState(initialFile?.basename || initialTitle);
+  const [isOpen, setIsOpen] = useState(false);
+  const [activeIndex, setActiveIndex] = useState(0);
+  const [isFocused, setIsFocused] = useState(false);
+  const [searchResults, setSearchResults] = useState<TFile[]>([]);
+  const [isSearching, setIsSearching] = useState(false);
+  const queryEngine = useRef(new QueryEngine(plugin.app));
+  const titleInputRef = useRef<HTMLInputElement>(null);
+  const popoverRef = useRef<HTMLDivElement>(null);
+  const menuRef = useRef<HTMLUListElement>(null);
+  const debounceTimeoutRef = useRef<number | null>(null);
+
+  // Search for nodes when query changes (only in create mode)
+  useEffect(() => {
+    if (isEditMode) {
+      setSearchResults([]);
+      return;
+    }
+    const searchQuery = query.trim();
+    if (searchQuery.length < 2) {
+      setSearchResults([]);
+      return;
+    }
+
+    if (debounceTimeoutRef.current) {
+      clearTimeout(debounceTimeoutRef.current);
+    }
+
+    setIsSearching(true);
+    debounceTimeoutRef.current = window.setTimeout(() => {
+      void (async () => {
+        try {
+          const results = selectedNodeType
+            ? await queryEngine.current.searchDiscourseNodesByTitle(
+                searchQuery,
+                selectedNodeType.id,
+              )
+            : await queryEngine.current.searchDiscourseNodesByTitle(
+                searchQuery,
+              );
+          setSearchResults(results);
+        } catch (error) {
+          console.error("Error searching nodes:", error);
+          setSearchResults([]);
+        } finally {
+          setIsSearching(false);
+        }
+      })();
+    }, 250);
+    return () => {
+      if (debounceTimeoutRef.current) {
+        clearTimeout(debounceTimeoutRef.current);
+      }
+    };
+  }, [query, selectedNodeType, isEditMode]);
+
+  useEffect(() => {
+    if (!selectedExistingNode) {
+      setTitle(query);
+    }
+  }, [query, selectedExistingNode]);
```

Comment:
Is this necessary?  I worry about race conditions/duplicate work.
Can this be handled where query is set?